### ISSUE TITLE
CustomElements Event propagation broken

### ISSUE THREAD

**Issue by martok** [martok/palefill#30]
Among other things, soft navigation in github doesn't work. This is because the event retargeting shim is extremely limited and not spec compliant.

A proper implementation could look something like https://github.com/salesforce/lwc/blob/master/packages/%40lwc/synthetic-shadow/src/faux-shadow/events.ts

**Comment by martok** [comment]
This one special case is fixed, but in general custom elements have some sort deeper issue where they sometimes steal events from their children and sometimes their children bubble events to the wrong level.

**Comment by martok** [comment]
> The reason why SeaHOH implemented this "button.onclick" special case is that the event doesn't get dispatched to the custom element inside the button. Almost as if they are not considered valid event targets?

This is indeed exactly the problem with the "Copy SHA" button. It is not the `<button>` that has the event attached but the `<clipboard-copy>` CE. The CE is then made big enough to cover the entire button, so it "looks like" the user presses a button.
Easy test: there are a few pixels on the left where the button is larger than the CE where Chromium doesn't copy anything, same as Pale Moon. I've set a reddish background for the CE here so it's visible:
![image](https://user-images.githubusercontent.com/109197/202938199-92ea4188-880e-434f-ab4d-8f3a3a7edd01.png)

With keyboard navigation, same thing: the button doesn't do anything, the CE must be focused. Which interestingly works in Pale Moon!
Now, the question is: why does the customElement not receive mouse events. It doesn' even capture the DevTools inspector context menu.

**Comment by SeaHOH** [comment]
This must be a bug of `<button>`, other custom elements can receive mouse events.

**Comment by martok** [comment]
Wouldn't be the first, the thing that confuses me is why it's only mouse events and keyboard works.

Although... the hint bug that @Vangelis66 found [on the other issue](https://github.com/martok/palefill/issues/60#issuecomment-1312823509) could be a hit test issue as well. Assuming an element inside `<button>` doesn't take part in hit tests, it's hint wouldn't be shown as well as clicks not delivered.

**Comment by SeaHOH** [comment]
Polyfill is weak, there is nothing more we can do, just have to wait Moonchild to fix it.

**Comment by martok** [comment]
... or contribute to Pale Moon. Which I intend to do with these (assuming I can still build it), the links were mostly notes to myself.

Other than that, the polyfill I've been working on over the weekend should fix the remaining event issues. Just need to async-ify it a bit and be a bit less paranoid about DOM updates, currently it is way too slow.

**Comment by martok** [comment]
> currently it is way too slow.

Show of hands: is ~1.5s slower (using this issue page as test case) compared to Palefill 1.23 acceptable?

For me, that changes DOMContentLoaded (using Network panel of devtools) from 10s to 11.5s, which is overall barely noticeable difference. Of that, the first 5s are HTML parsing, the first JS task gets processed at around 6s and the last shadow operation is at about 9s.

**Comment by garoto** [comment]
Is this issue related to the timestamping one and to your PR in palemoon's repo that Vangelis66 mentioned some hours ago? Because I have no idea what `soft navigation` is, but otherwise, as long the 1.5s increase in rendering time is the penalty for having a non-broken GH, then I raise my hands I guess.

**Comment by Vangelis66** [comment]
Thanks @martok for your efforts, all the more since you're pressed for spare time...

If "soft-navigation" is to come back, will (currently open) #53 be addressed?
And, are you talking about [this](https://gist.github.com/martok/d675bead18a98a087d84f506d327b504)? 
My primary concern is that #60 is somehow addressed, because the **lack or relative timestamps** slows down my workflow 😞 ...

As for **longer page-rendering times**, I have mixed feelings... Obviously, the times measured also depend on the **underlying hardware** (CPU+GPU+RAM), not only on the code running; (I guess) people on older machines (and older OSes using the UXP forks) will be affected more than, say, a user on an i7+16GB of RAM on Win10x64...

Despite the above reservation, you have my blessing to go forward with whatever is deemed necessary **to get GitHub fully working again** (until MS, on a whim, break it again for us 😠 ) ...

@garoto 

> I have no idea what soft navigation is

From "https://github.com/WICG/soft-navigations#readme":

> "Soft navigations" are JS-driven same-document navigations that are using the history API or the new Navigation API, triggered by a user gesture and modify the DOM, modifying the previous content, as well as the URL displayed to the user.

and from "https://www.infoq.com/news/2015/11/performance-web-applications/": 

> Soft navigations are not real navigations, that is, you can change a browser's URL without really navigating and thus, without triggering onload events. And the browsers don't have an API that informs if and when all the resources - i.e.: images, scripts, CSS, videos - have been downloaded.

**TL;DR**: Say you visit 
https://github.com/martok/palefill
in Google Chrome/Mozilla Firefox; by default, you should land in the `<>Code` tab; when you click, e.g., the `Insights` tab (within the same GH page), this is loaded via **soft-navigation**; the browser's URL bar will change to "https://github.com/martok/palefill/pulse", while in the process of doing so GitHub will briefly display **a thin blue progress line** (from left to right) at the top of the page...

> Is this issue related to ... your PR in palemoon's repo that Vangelis66 mentioned some hours ago?

The [UXP PR](https://repo.palemoon.org/MoonchildProductions/UXP/pulls/2043) I mentioned in the [other issue](https://github.com/martok/palefill/issues/60) intends to remedy the bug discussed [here](https://github.com/martok/palefill/issues/60#issuecomment-1312578916) and [here](https://github.com/martok/palefill/issues/60#issuecomment-1312823509) ...

**Comment by garoto** [comment]
Alright, cheers Vangelis66 for the detailed explanation.

**Comment by martok** [comment]
> Thanks @martok for your efforts, all the more since you're pressed for spare time...

It's getting better (or my time management is), and today I've got notification that a grant proposal deadline I expected for April is going to be in June :smile: 

> If "soft-navigation" is to come back, will (currently open) #53 be addressed?

I'm not sure. Something there causes massive memory leaks (see #50), so it doesn't seem worth it. We'll see.

>  And, are you talking about [this](https://gist.github.com/martok/d675bead18a98a087d84f506d327b504)?

Yep. Or rather, a much more refined version of that concept (turns out it can't work like that) that will also be released as a separate project. Because who doesn't want yet another ShadowDOM polyfill in their life :monocle_face: 

> My primary concern is that #60 is somehow addressed, because the **lack of relative timestamps** slows down my workflow disappointed ...

That, and hopefully a few others. Haven't had a release in a while and stuff piles up... should be out before the end of the week, I also want to retest a few old bugs and see if we can get away with less weird patches now.

> As for **longer page-rendering times**, I have mixed feelings...

Same, but honestly: if you want a fast experience, don't just Pale Moon. I have no idea why that is, but complex documents even without any scripting are so much slower than in other browsers, it's almost funny given how I discovered PM waaaay back when it was "just" a more optimized build of Firefox.

> The [UXP PR](https://repo.palemoon.org/MoonchildProductions/UXP/pulls/2043) I mentioned in the [other issue](https://github.com/martok/palefill/issues/60) intends to remedy the bug discussed [here](https://github.com/martok/palefill/issues/60#issuecomment-1312578916) and [here](https://github.com/martok/palefill/issues/60#issuecomment-1312823509) ...

That just got merged an hour ago :)

**Comment by martok** [comment]
Well, we know where the remaining issue comes from and I've likely fixed everything regarding element propagation that can be fixed. So I'll close this and leave issue #64 to track that one problem.
